src/main/software/amazon/event/ruler/GenericMachine.java

```diff
@@ -533,35 +534,34 @@ private boolean addStep(final NameState state,
         // for each pattern, we'll provisionally add it to the BMC, which may already have it.  Pass the states
         //  list in in case the BMC doesn't already have a next-step for this pattern and needs to make a new one
         //
-        NameState lastNextState = null;
-        Set<NameState> nameStates = new HashSet<>();
+        NameState nextNameState = state.getNextNameState(key);
+        if (nextNameState == null) {
+            nextNameState = new NameState();
+            state.addNextNameState(key, nextNameState);
+        }
         for (Patterns pattern : patterns.get(key)) {
             if (isNamePattern(pattern)) {
                 assert nameMatcher != null;
-                final NameState nameStateForSupplier = lastNextState == null ? new NameState() : lastNextState;
-                lastNextState = nameMatcher.addPattern(pattern, () -> nameStateForSupplier);
+                final NameState nameStateForSupplier = nextNameState;
+                nameMatcher.addPattern(pattern, () -> nameStateForSupplier);
             } else {
-                assert byteMachine != null;
-                lastNextState = byteMachine.addPattern(pattern, lastNextState);
+                byteMachine.addPattern(pattern, nextNameState);
             }
-            nameStates.add(lastNextState);
         }
-        nameStatesForEachKey.add(nameStates);
+        nameStateForEachKey.add(nextNameState);
 
         // Determine if we are adding a new rule or not. If we are not yet at the terminal key, go deeper recursively.
         // If we are at the terminal key, check if the NameState already contains a terminal sub-rule for the provided
         // rule name and patterns, and propagate this result up the recursion stack.
         boolean isRuleNew = false;
         final int nextKeyIndex = keyIndex + 1;
         boolean isTerminal = nextKeyIndex == keys.size();
-        for (NameState nameState : nameStates) {
-            if (!isTerminal) {
-                isRuleNew = addStep(nameState, keys, nextKeyIndex, patterns, ruleName, addedKeys, nameStatesForEachKey);
-            } else {
-                for (Patterns pattern : patterns.get(key)) {
-                    if (!nameState.containsTerminalSubRule(ruleName, pattern)) {
-                        return true;
-                    }
+        if (!isTerminal) {
+            isRuleNew = addStep(nextNameState, keys, nextKeyIndex, patterns, ruleName, addedKeys, nameStateForEachKey);
+        } else {
+            for (Patterns pattern : patterns.get(key)) {
+                if (!nextNameState.containsTerminalSubRule(ruleName, pattern)) {
```

Comment:
I see, agree with your analyze, 

>We need to take the intersection between the candidates and the terminal sub-rules. How do you take an intersection in O(1) time?

It looks like O(1) is impossible and it has to be O(n) as trade-off, n is min(`terminalSubRules.size` and `candidateSubRuleIds.size`), candidateSubRuleIds is from possible sub rules  and `terminalSubRules.size` o its size is bound to number of matches, with different situation its size might be different:
e.g.
```
        String rule1 = "{ \"a\" : [ 1 ], \"b\": [%d] }";
        for (int i = 0; i<50; i++) {
            machine.addRule("rule1", String.format(rule1,i));
        }
```
vs.
```
        String rule1 = "{ \"a\" : [ %d ], \"b\":  [ 1 ] }";
        for (int i = 0; i<50; i++) {
            machine.addRule("rule1", String.format(rule1,i));
        }
```
so, I am thinking the possible improvement to judge size and loop set with smaller size as improvement.
```
for(set with smaller size) {  // O(n_smaller)
   set with larger size.contains(x); // O(1)
} => O(n_smaller)
``` 
just my 2 cents.

maybe it is good to highlight user to try not create to many rules with similar pattern serial to avoid perf penalty for latency sensitive scenarios.

Reply:
Long, your insight never fails to amaze me. Last week, I heard from a highly-latency-sensitive team with a performance test that exercised this scenario: a large machine with many similar rules. This performance test showed a latency regression. I went through and made a handful of performance improvements, including the smaller set / larger set improvement you described above. I also reverted the improvement I made to further re-use NameState (as long as keys are the same). Although that improvement further reduced memory use, it increased runtime due to more sub-rules contained in each NameState. At this point, the runtime is only slightly increased (in the case of many similar repeated rules), but memory use is largely decreased. I think it's a good trade-off and I'd like to settle here for now. I'll push that revert and the other performance improvements to this PR in the next couple days.

I could see a future improvement by allowing a config to indicate how much to optimize for memory and how much to optimize for latency. It would certainly add more complexity though.

Reply:
I've pushed the revert and other performance improvements. Hopefully these are the last changes before merging into main.

Reply:
Thanks for the update Shawn, really happy to see you have found that team, experienced and improved the performance. Agree it is a good trade-off and big win to use a slight run-time increase to gain large memory saving, it's an awesome delivery.